packages/binding-http/src/http-server-middleware.ts

```diff
@@ -0,0 +1,83 @@
+/********************************************************************************
+ * Copyright (c) 2023 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the W3C Software Notice and
+ * Document License (2015-05-13) which is available at
+ * https://www.w3.org/Consortium/Legal/2015/copyright-software-and-document.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR W3C-20150513
+ ********************************************************************************/
+
+/**
+ * HTTP Middleware for the HTTP Server
+ */
+
+import { createLoggers } from "@node-wot/core";
+import * as http from "http";
+
+const { debug } = createLoggers("binding-http", "http-server-middleware");
+
+/**
+ * A function that handles a request and passes it to the WoT Servient.
+ * See {@link HttpMiddleware} for an example.
+ * @param req The HTTP request.
+ * @param res The HTTP response.
+ * @param next Call this function to pass the request to the WoT Servient.
+ */
+export type MiddlewareRequestHandler = (
+    req: http.IncomingMessage,
+    res: http.ServerResponse,
+    next: () => void
+) => Promise<void>;
+
+/**
+ * A middleware for the HTTP server, which can be used to intercept requests before they are handled by the WoT Servient.
+ *
+ * Example:
+ * ```javascript
+ * import { Servient } from "@node-wot/core";
+ * import { HttpMiddleware, HttpServer } from "@node-wot/binding-http";
+ *
+ * const servient = new Servient();
+ *
+ * const middleware = new HttpMiddleware(async (req, res, next) => {
+ *     // For example, reject requests in which the X-Custom-Header header is missing
+*      // by replying with 400 Bad Request
+ *     if (!req.headers["x-custom-header"]) {
+ *         res.statusCode = 400;
+ *         res.end("Bad Request");
+ *         return;
+ *     }
+ *     // Pass all other requests to the WoT Servient
+ *     next();
+ * });
+
+ * const httpServer = new HttpServer({
+ *     port,
+ *     middleware,
+ * });
+ *
+ * servient.addServer(httpServer);
+ *
+ * servient.start().then(async (WoT) => {
+ *    // ...
+ * });
+ * ```
+ */
+export default class HttpMiddleware {
+    public handler: MiddlewareRequestHandler;
+
+    constructor(handler: MiddlewareRequestHandler) {
+        this.handler = handler;
+    }
+
+    public handleRequest(req: http.IncomingMessage, res: http.ServerResponse, next: () => void): Promise<void> {
+        debug("Hit middleware");
+        return this.handler(req, res, next);
+    }
+}
```

Comment:
Can't we just get rid of this class and use directly the `MiddlewareRequestHandler`. I don't see the benefit of having this container class other than having a debug message. 🤔

Reply:
Maybe, it would be useful to have a class that takes as input a list of `MiddlewareRequestHandler`s and configure the next function accordingly? (just got this requirement from my local refactoring, but feel free to add it later)

Reply:
> Can't we just get rid of this class and use directly the MiddlewareRequestHandler. I don't see the benefit of having this container class other than having a debug message.

The class was meant more as a base class that can a developer can extend to handle more complex logic, like keeping a "state" between on request and the other, or any other logic that comes to mind.

But I agree the `MiddlewareRequestHandler` function is what gets executed inside the HttpServer, so we can get rid of the class. Just let me know how to proceed.

> Maybe, it would be useful to have a class that takes as input a list of `MiddlewareRequestHandler`s and configure the next function accordingly? (just got this requirement from my local refactoring, but feel free to add it later)

How would you configure the `next` function based on the `MiddlewareRequestHandler`? I didn't really get it :thinking:

Reply:
> Maybe, it would be useful to have a class that takes as input a list of `MiddlewareRequestHandler`s and configure the next function accordingly? (just got this requirement from my local refactoring, but feel free to add it later)

Maybe the type of the `middleware` parameter in the `HttpServer`'s constructor could also be changed to `HttpMiddleware | HttpMiddleware[]` to cover that? Although things might indeed become a bit complicated in the array case when it comes to the `next` function.

Maybe an alternative could be to let users pass in functions of type

```ts
export type AlternativeMiddlewareRequestHandler = (
    req: http.IncomingMessage,
    res: http.ServerResponse,
    abort: () => void,
) => Promise<void>;
```

and then the server does the chaining internally. In this case, I replaced `next()` with an `abort()` function which would allow users to stop the processing – the next middleware would be called by the server instead if no `abort()` happened before.

Do you think this could work? Or would having a `next()` function instead give users more control?

Reply:
I was thinking of something like this:
```
let middlewares: MiddlewareRequestHandler[] // initialized somewhere
const first = middlewares.shift()
const entries = middleware.entries()

first(req,res, () => entries.next().value[1]()) // needs to check also if we are done but removed for giving the idea
```

this way the middleware can control whether to pass to the next function or not, without introducing the abort. Going back to the PR itself @Luca8991 I would remove HTTPMiddleware class and discuss further improvements in a future PR.

Reply:
@relu91 I've pushed the commit that removes the `HTTPMiddleware` class